### ISSUE TITLE
Device becomes unavailable

### ISSUE THREAD

**Issue by gminadak** [rospogrigio/localtuya#1268]
## The problem

I have an "Aubess Smart Socket 20A/EM" it's the only tuya device currently connected to HA.
The problem is that from time to time the device becomes unavailable at random time's through the day.
If I reload the integration the device becomes available again for some time.

## Environment

- Localtuya version: v5.0.0
- Home Assistant Core version: 2023.1.7
- [] Does the device work using the Home Assistant Tuya Cloud component ? 
- [] Does the device work using the Tinytuya (https://github.com/jasonacox/tinytuya) command line tool ? 
- [] Was the device working with earlier versions of localtuya ? Which one? 
- [X] Are you using the Tuya/SmartLife App in parallel ? 
*Just installed it on my phone never using it for actions.

## Steps to reproduce

1. Just installed local tuya by following the instructions.

## DP dump
```
INFO:localtuya:localtuya version 1.0.0
INFO:localtuya:Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
INFO:localtuya:Using pytuya version '10.0.0'
INFO:localtuya:Detecting list of available DPS of device ********************** [192.168.1.40], protocol 3.3.
DEBUG:localtuya.pytuya:[bfa...zgv] Sending command 10 (device type: type_0a)
DEBUG:localtuya.pytuya:[bfa...zgv] Sending payload: b'{"gwId":"**********************","devId":"**********************","uid":"**********************","t":"1674991543"}'
DEBUG:localtuya.pytuya:[bfa...zgv] Command 10 waiting for sequence number 1
DEBUG:localtuya.pytuya:[bfa...zgv] Dispatching message CMD 10 TuyaMessage(seqno=1, cmd=10, retcode=1, payload=b'6\xd5\xc6$\xb2]\xf6\x95f*\xf1r\xbf\xda\xca\xa3\xb8@D\x8b\x17\xd5d\x96\x00\r\xc1\x87\x00T\x93w', crc=3611199455, crc_good=True)
DEBUG:localtuya.pytuya:[bfa...zgv] incomplete payload=b'6\xd5\xc6$\xb2]\xf6\x95f*\xf1r\xbf\xda\xca\xa3\xb8@D\x8b\x17\xd5d\x96\x00\r\xc1\x87\x00T\x93w' (len:32)
DEBUG:localtuya.pytuya:[bfa...zgv] ERROR Unexpected Payload from Device - 904 - payload: null
AVAILABLE DPS ARE [{}]
INFO:localtuya:COMPLETE response from device ********************** [192.168.1.40].

**** deviceInfo returned OK ****

TuyaDebug (Tuya DPs dump) [1.0.0]

Device ********************** at 192.168.1.40 key ******************************** protocol 3.3 dev_type type_0a:
```

*I'm not sure if this is what you need here
Switch 1          1
Relay status      38
Overcharge switch 39
Light mode        40
Countdown 1       9
Child Lock	      41
Cycle Time	      42
Random Time	      42
Inching Switch	  44
Add Electricity   17
Current           18
Power             19
Voltage           20

## Provide Home Assistant traceback/logs

```
2023-01-29 12:43:16.248 ERROR (MainThread) [custom_components.localtuya.common] [bfa...zgv] Connect to 192.168.1.40 failed
Traceback (most recent call last):
  File "/config/custom_components/localtuya/common.py", line 187, in _make_connection
    self._interface = await pytuya.connect(
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 1163, in connect
    _, protocol = await loop.create_connection(
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 1064, in create_connection
    raise exceptions[0]
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 1049, in create_connection
    sock = await self._connect_sock(
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 960, in _connect_sock
    await self.sock_connect(sock, address)
  File "/usr/local/lib/python3.10/asyncio/selector_events.py", line 500, in sock_connect
    return await fut
  File "/usr/local/lib/python3.10/asyncio/selector_events.py", line 535, in _sock_connect_cb
    raise OSError(err, f'Connect call failed {address}')
TimeoutError: [Errno 110] Connect call failed ('192.168.1.40', 6668)```


## Additional information
I see that there are other open reports for this issue, my logs look different also I can't get the DP dump.

**Comment by escobarin3** [comment]
My Local Tuya integration has not been working for a while now... at least for the las 2 weeks. My devices become unavailable every now and then... it is becoming very unrealiable. This is a new thing obviosly because it worked fine for years....  :(

I cannot find anyting in my logs that indicates any possible issue... 
I cannot edit a device or reload the integration bacause all my devices go offline/unavailable... forcing me to restart HA... 
Any ideas?

My most recent log shows this: 

Logger: homeassistant.components.websocket_api.http.connection
Source: config_entries.py:1111
Integration: Home Assistant WebSocket API (documentation, issues)
First occurred: 8:50:46 PM (1 occurrences)
Last logged: 8:50:46 PM

[139952805691888] The config entry localtuya (localtuya) with entry_id c0423c56a1535434046c08917f78ee30 cannot be unloaded because it is not in a recoverable state (ConfigEntryState.FAILED_UNLOAD)
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 200, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1755, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1792, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 751, in admin_handler
    await result
  File "/config/custom_components/localtuya/__init__.py", line 84, in _handle_reload
    await asyncio.gather(*reload_tasks)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 1128, in async_reload
    unload_result = await self.async_unload(entry_id)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 1111, in async_unload
    raise OperationNotAllowed(
homeassistant.config_entries.OperationNotAllowed: The config entry localtuya (localtuya) with entry_id c0423c56a1535434046c08917f78ee30 cannot be unloaded because it is not in a recoverable state (ConfigEntryState.FAILED_UNLOAD)

**Comment by Ayo91** [comment]
Same issue at my side, I cannot connec to de devices:

`Unexpected exception
Traceback (most recent call last):
  File "/config/custom_components/localtuya/config_flow.py", line 581, in async_step_configure_device
    self.dps_strings = await validate_input(self.hass, user_input)
  File "/config/custom_components/localtuya/config_flow.py", line 240, in validate_input
    interface = await pytuya.connect(
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 1163, in connect
    _, protocol = await loop.create_connection(
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 1064, in create_connection
    raise exceptions[0]
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 1049, in create_connection
    sock = await self._connect_sock(
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 960, in _connect_sock
    await self.sock_connect(sock, address)
  File "/usr/local/lib/python3.10/asyncio/selector_events.py", line 500, in sock_connect
    return await fut
  File "/usr/local/lib/python3.10/asyncio/selector_events.py", line 535, in _sock_connect_cb
    raise OSError(err, f'Connect call failed {address}')
OSError: [Errno 113] Connect call failed ('192.168.1.185', 6668)`

**Comment by SDeGonge** [comment]
Same here. Reloading localtuya brings everything back but within 10 minutes one or more of my devices becomes unavailable.

**Comment by alexander250100** [comment]
Got the same problems, I have it for a couple of weeks now, sometimes reloading localtuya and rebooting router helps.

**Comment by alexander250100** [comment]
> Same here. Only solution I found is to downgrade to version 4.0.2.

Downgraded, looks likes it works for now, thanks for your suggestion.

**Comment by gminadak** [comment]
@hpommier Downgraded same problem after a couple of hours.
Reloading localtuya still work's with 4.0.2

**Comment by doddys** [comment]
downgrading to 4.0.2 proves to be more stable for my case as well

**Comment by PhilippeP62** [comment]
Same thing here , some devices goes missing for minutes or hours...

This error originated from a custom integration.

Logger: custom_components.localtuya.common
Source: custom_components/localtuya/pytuya/__init__.py:1163
Integration: LocalTuya (documentation, issues)
First occurred: February 1, 2023 at 22:48:31 (419 occurrences)
Last logged: 14:03:07

[bfb...ltk] Connect to 192.168.3.35 failed
[bf5...a4j] Connect to 192.168.3.28 failed
Traceback (most recent call last):
  File "/config/custom_components/localtuya/common.py", line 187, in _make_connection
    self._interface = await pytuya.connect(
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 1163, in connect
    _, protocol = await loop.create_connection(
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 1064, in create_connection
    raise exceptions[0]
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 1049, in create_connection
    sock = await self._connect_sock(
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 960, in _connect_sock
    await self.sock_connect(sock, address)
  File "/usr/local/lib/python3.10/asyncio/selector_events.py", line 500, in sock_connect
    return await fut
  File "/usr/local/lib/python3.10/asyncio/selector_events.py", line 535, in _sock_connect_cb
    raise OSError(err, f'Connect call failed {address}')
TimeoutError: [Errno 110] Connect call failed ('192.168.3.35', 6668)


.35 is a light dimmer controller and is currently unavailable.
.28  is a temp/humidity sensor.

**Comment by alexander250100** [comment]
try version 4.0.2, my problems are solved.

**Comment by SDeGonge** [comment]
Hmmm, how do I revert to 402?

**Comment by alexander250100** [comment]
download 4.02 from this site and overwrite files with file editor in HA, restart system

**Comment by PhilippeP62** [comment]
Gone back to 402 doesn't seems to do anything , will see tomorrow if it comes back

This error originated from a custom integration.

Logger: custom_components.localtuya.common
Source: custom_components/localtuya/pytuya/__init__.py:669
Integration: LocalTuya (documentation, issues)
First occurred: 21:32:51 (1 occurrences)
Last logged: 21:32:51

[bfb...ltk] Connect to 192.168.3.35 failed
Traceback (most recent call last):
  File "/config/custom_components/localtuya/common.py", line 160, in _make_connection
    self._interface = await pytuya.connect(
  File "/config/custom_components/localtuya/pytuya/__init__.py", line 669, in connect
    _, protocol = await loop.create_connection(
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 1064, in create_connection
    raise exceptions[0]
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 1049, in create_connection
    sock = await self._connect_sock(
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 960, in _connect_sock
    await self.sock_connect(sock, address)
  File "/usr/local/lib/python3.10/asyncio/selector_events.py", line 500, in sock_connect
    return await fut
  File "/usr/local/lib/python3.10/asyncio/selector_events.py", line 535, in _sock_connect_cb
    raise OSError(err, f'Connect call failed {address}')
TimeoutError: [Errno 110] Connect call failed ('192.168.3.35', 6668)

**Comment by Bazooka2304** [comment]
I have the same issue with 5.0.0. I had deleted Local Tuyal and configured all devices new but after a while all the devices disappeared again.

**Comment by The-element** [comment]
Revert to 4.1.1. and setup an automation to reload the integration every night seems to solve it for me.

**Comment by PhilippeP62** [comment]
So after going back to 5.0 again , it was still unavailable this morning but it currently works .

**Comment by paulius2k** [comment]
I am having same issues since ~2023.01.15.

Reverting to 4.1.1 did not help - still devices go unavailable.

I will try out 4.0.2 as was suggested here.

**Comment by Ayo91** [comment]
Im removing all my wifi tuya devices and try to change by tuya zigbee ones with zha compatibility: https://zigbee.blakadder.com/zha.html

Maybe it helps!

**Comment by paulius2k** [comment]
Update after trying v4.0.2 - my devices are half-functional (I have seen same behaviour with v5.0.0 and v 4.1.1): 
- turning Ledvance Smart Plug on/off does not work, I get error:
```
14:17:16.517 WARNING (MainThread) [homeassistant.helpers.service] Unable to find referenced entities switch.ledvance_smart_plug or it is/they are currently not available
```
- when I turn on/off the switch using it's physical button, I do get proper status update to my HA:
```
2023-02-06 14:23:37.561 DEBUG (MainThread) [custom_components.localtuya.pytuya] [011...9d7] Got status update
2023-02-06 14:23:37.564 DEBUG (MainThread) [custom_components.localtuya.pytuya] [011...9d7] Decrypted payload: {"devId":"xx","dps":{"18":0,"19":0},"t":1675686216}
```

- Heartbeat also comes through:
```
2023-02-06 14:23:39.877 DEBUG (MainThread) [custom_components.localtuya.pytuya] [011...9d7] Sending command heartbeat (device type: type_0a)
2023-02-06 14:23:39.877 DEBUG (MainThread) [custom_components.localtuya.pytuya] [011...9d7] Send payload: b'{}'
2023-02-06 14:23:39.880 DEBUG (MainThread) [custom_components.localtuya.pytuya] [011...9d7] Waiting for sequence number -100
2023-02-06 14:23:39.909 DEBUG (MainThread) [custom_components.localtuya.pytuya] [011...9d7] Dispatching message TuyaMessage(seqno=0, cmd=9, retcode=0, payload=b'', crc=2958142211)
2023-02-06 14:23:39.909 DEBUG (MainThread) [custom_components.localtuya.pytuya] [011...9d7] Got heartbeat response
2023-02-06 14:23:39.910 DEBUG (MainThread) [custom_components.localtuya.pytuya] [011...9d7] Decrypted payload: {}
```

**Comment by AlphaZuluNRW** [comment]
I've the same problem. I'll try version 4.0.2 too.

**Comment by SDeGonge** [comment]
I've reverted to 4.0.2 and everything seems to be stable now.

**Comment by oat1** [comment]
Did someone try on new beta ?
<img width="418" alt="image" src="https://user-images.githubusercontent.com/59432560/217547930-7168dacf-56b3-44dd-94b5-07fe42355301.png">

**Comment by escobarin3** [comment]
I did as explained in the following comment from another issue and everything has been fine since then: 

Turns out this was indeed the problem, and the easiest workaround (and probably friendlier to future updates) for now is to:

open /config/.storage/core.config_entries in a text editor
search for the device entry
look for the "scan_interval" key (if it doesn't exist, use config flow to edit the device and set this to an integer)
remove the quotes around the value, i.e. "15" should be changed to 15

taken from: 
https://github.com/rospogrigio/localtuya/issues/1253#issuecomment-1412069911

**Comment by RaRo1980** [comment]
has the same Problems and with 5.1.0-beta0 looks like is stable and is working

**Comment by paulius2k** [comment]
`v5.1.0-beta0` seems to be working fine for the last day.

**Comment by przemas11** [comment]
I had the same problem with disconnecting. I have installed 5.1.0-beta0 and I had 8 disconnects today but after each disconnect the connection came back so it is looking good.

**Comment by gminadak** [comment]
I can confirm that with 5.1.0-beta0 the device always reconnects after a while.

![image](https://user-images.githubusercontent.com/45330594/218020961-939c370b-2096-40e4-85ea-779f278b05c5.png)

```
2023-02-10 03:23:15.003 WARNING (MainThread) [custom_components.localtuya.pytuya] [bfa...zgv] Command 9 timed out waiting for sequence number -100
2023-02-10 03:23:15.006 WARNING (MainThread) [custom_components.localtuya.common] [bfa...zgv] Disconnected - waiting for discovery broadcast
2023-02-10 03:25:29.431 WARNING (MainThread) [custom_components.localtuya.common] [bfa...zgv] Failed to connect to 192.168.1.40: [Errno 110] Connect call failed ('192.168.1.40', 6668)
2023-02-10 03:27:44.599 WARNING (MainThread) [custom_components.localtuya.common] [bfa...zgv] Failed to connect to 192.168.1.40: [Errno 110] Connect call failed ('192.168.1.40', 6668)
2023-02-10 03:29:59.769 WARNING (MainThread) [custom_components.localtuya.common] [bfa...zgv] Failed to connect to 192.168.1.40: [Errno 110] Connect call failed ('192.168.1.40', 6668)
```

**Comment by aritorhamo** [comment]
> open /config/.storage/core.config_entries in a text editor
search for the device entry
look for the "scan_interval" key (if it doesn't exist, use config flow to edit the device and set this to an integer)
remove the quotes around the value, i.e. "15" should be changed to 15

I don't have "scan_interval" in my core.config_entries file. I'm not an IT engineer (as most people here seem to be...), so I have to ask, do you with "conflig flow" refer to some of the files named "config_flow.py"? If yes, are those scripts that I should run on the terminal, and which one of them should I use? Or would it be easier to upgrade to a later version LocalTuya (my current version is 5.0.0).

My problem is similar to what others here are experiencing: my Ledvance lights keep becoming "unavailable". Starting them in the Ledvance app helps temporarily, but the lights keep disappearing. Currently I have only one of my 15 lights left working.

**Comment by The-element** [comment]
Tried the 5.1.0-beta0 and it doesn't solve it for me. So added an automation to reload the integration once every hour. Still doesn't work. Now all my devices has been unavailable for 5 hours. Again.
![image](https://user-images.githubusercontent.com/59605032/218402220-da1f23d5-9b68-476e-a991-eb60e14732aa.png)

**Comment by gminadak** [comment]
I have now completely remove localtuya so I'm closing this issue.
As I can see there are other open issues with this problem if the developer find a solution it will be posted there.
Thank you all for your contribution.
